### PR TITLE
Port REACT_MODULE() from react-native-windows to react-native-macos (3/4)

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

This is the second change for porting REACT_MODULE() from react-native-windows
It enables user to run a turbo module in C++ like this
```c++
REACT_MODULE(MyModule)
struct MyModule
{
    REACT_METHOD(MyMethod)
    void MyMethod(string a, int b, ReactPromise<double> result) noexcept
    {
        result.Reject("Not implemented");
    }
};
```

- This change moves some files from react-native-windows, they defines `REACT_MODULE` with a proper reflection system
- All copied files will eventually be downloaded from packages, the duplication is just a temporary solution for developing
- The turbo module written using `REACT_MODULE` is executable, and the threading supported is added to complete the `Screenshot/Snapshot` test page.

## For Code Review

The original code is deeply coupled with WinRT,
I have no choice but to implement involved WinRT API in macOS,
here is two files that do this job

- **RNTester/RNTester-macOS/TurboModuleCxx/winrt/Windows.Foundation.h**
- **RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h**

For other files in **RNTester/RNTester-macOS/TurboModuleCxx**,
they are copied from react-native-windows,
which mean I am not going to fix issues about these files in this pull request
(because such changes are expected to be submitted to react-native-windows),
except for code in `#ifdef __APPLE__`.

As what is mentioned above,
eventually these files will be downloaded from a package,
so after a few pull requests,
they will be deleted from this repo.

Having duplicated files here is a temporary solution for development convenience.

## About Future Pull Requests

`Screenshot/Snapshot` is running on `REACT_MODULE` now.
I will put files in cocoapods in the next pull request.

## Changelog

[macOS] [Fixed] - Port REACT_MODULE() from react-native-windows to react-native-macos (3/4)

## Test Plan

- Manually tested

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-macos/pull/608)

### CODE REVIEW

File: RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.MacOS.mm [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  bool HasThreadAccess() const noexcept
+  {
+    VerifyElseCrash(false);
+  }
+  void Post(const ReactDispatcherCallback& callback) noexcept
+  {
+    auto copy = callback;
+    dispatch_async(dispatch_get_main_queue(), ^{
+      copy();
+    });
```
Comment (alloy) [inline RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.MacOS.mm:18]: Btw, if helpful, I’m pretty sure that you can pass C++ lambda’s to libdispatch APIs.

File: RNTester/NativeModuleExample/ScreenshotMacOS.mm [hunk 2]
```diff
@@ -9,11 +9,10 @@
 #import <React/RCTUIManager.h>
 #import <React/RCTUtils.h>
 #import <ReactCommon/RCTTurboModuleManager.h>
-#import <ReactCommon/TurboModuleUtils.h>
 #import <NativeModules.h>
 #import <TurboModulesProvider.h>
 
-static NSImage* TakeScreenshot()
+static NSImage* TakeScreenshotAsImage()
```
Comment (alloy) [inline RNTester/NativeModuleExample/ScreenshotMacOS.mm:15]: I only just now notice this, and it’s been a while since I’ve had to think about this, but I believe the Core Graphics image [created here below](https://github.com/microsoft/react-native-macos/blob/57015042c4d1daeb90ee15ed233ede865642cf67/RNTester/NativeModuleExample/ScreenshotMacOS.mm#L31) might be leaking.

In principle the memory rule is that memory allocated with an API that has the word “Create” or “Copy” needs to be released. With objc objects these are nowadays managed by ARC, but the objects returned from these C APIs are not automatically included unless you transfer ownership to ARC. (It might be best to verify if ARC is inserting the release statement by looking at the compilation result.)

In short, I think the code should be:

```objc
  CGImageRef windowImage = CGWindowListCreateImage(…);
  NSImage* image = [[NSImage alloc] initWithCGImage:windowImage size:[keyWindow frame].size];
  CGImageRelease(windowImage);
```

Perhaps @tom-un’s or @HeyImChris’ knowledge on the status quo of these rules is more up-to-date.

File: RNTester/NativeModuleExample/ScreenshotMacOS.mm [hunk 3]
```diff
@@ -137,77 +79,37 @@ void TakeScreenshot(
...
-    VerifyElseCrash(false);
+    _reactContext.UIDispatcher().Post([this, result]
+    {
+      NSImage* screenshotImage = TakeScreenshotAsImage();
+      _reactContext.JSDispatcher().Post([screenshotImage, result]()
+      {
+        NSString* tempFilePath = SaveScreenshotToTempFile(screenshotImage);
+        result.Resolve([tempFilePath UTF8String]);
+      });
+    });
```
Comment (alloy) [inline RNTester/NativeModuleExample/ScreenshotMacOS.mm:90]: Short and sweet 👌

**Review by alloy — APPROVED** [review]
👍

File: RNTester/NativeModuleExample/ScreenshotMacOS.mm (hunk 2 above)
Comment (tom-un) [reply RNTester/NativeModuleExample/ScreenshotMacOS.mm:15]: Yes, good catch: the CGImageRef needs to be release via CGImageRelease or CFRelease

**Review by tom-un — DISMISSED** [review]

File: RNTester/NativeModuleExample/ScreenshotMacOS.mm (hunk 2 above)
Comment (ZihanChen-MSFT) [reply RNTester/NativeModuleExample/ScreenshotMacOS.mm:15]: Fixed

Comment (alloy) [reply RNTester/NativeModuleExample/ScreenshotMacOS.mm:15]: All objc objects are managed by ARC, all Core Foundation objects are not (by default).

These are probably good places to start and get up to speed fast:
* https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-CJBEJBHH
* https://developer.apple.com/library/archive/releasenotes/ObjectiveC/RN-TransitioningToARC/Introduction/Introduction.html#//apple_ref/doc/uid/TP40011226-CH1-SW1

**Comment by alloy** [comment]
The failing CI is caused by an ongoing CocoaPods CDN server issue, the tests were passing before the memory leak fix.
